lib/fizzy/parser_expr.cpp

```diff
@@ -103,16 +133,16 @@ void update_operand_stack(const ControlFrame& frame, Stack<ValType>& operand_sta
 
         const auto expected_type = *it;
         const auto actual_type = operand_stack.pop();
-        if (actual_type != expected_type)
+        if (!type_matches(actual_type, expected_type))
             throw validation_error{"type mismatch"};
     }
     // Push output values even if frame is unreachable.
     for (const auto output_type : outputs)
-        operand_stack.push(output_type);
+        operand_stack.push(from_valtype(output_type));
 }
 
-inline void drop_operand(
-    const ControlFrame& frame, Stack<ValType>& operand_stack, std::optional<ValType> expected_type)
+inline void drop_operand(const ControlFrame& frame, Stack<OperandStackType>& operand_stack,
+    OperandStackType expected_type)
```

Comment:
Wouldn't it be actually easier codewise to keep using `optional<ValType>`, not introducing `OperandStackType`, and comparing that way?

Reply:
I did it, too, at some point. These two variants are quite close, but enum one is perhaps a bit more concise and closer to Validation Algorithm.

Also enum way guarantees stack operands to be 1 byte each, without "has_value" flag overhead in `optional`.

(Also "Unknown" maybe better expresses the notion of "value with unknown/opaque type" than empty `optional`, which is rather supposed to mean `no value/no type`)

Reply:
We still use a mixture of `optional<ValType>` and `OperandStackType`. Perhaps we should aim to only use `OperandStackType` where possible?

Reply:
I don't think so, the remaining uses of `optional<ValType>` are all for block types, where empty one really means "no result value" (and so for example at jumps this means nothing is kept on stack; keeping `Unknown` would be incorrect)